lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,196 @@
+  auto *fnTy = fn.getFunctionType();
+  argTys.append(fnTy->params().begin(), fnTy->params().end());
+
+  auto *newFn = mutateFunctionArguments(fn, Type::getVoidTy(context), argTys, fn.getAttributes());
+
+  fn.replaceAllUsesWith(newFn);
+  for (unsigned idx = 0; idx < fn.arg_size(); idx++) {
+    Value *oldArg = fn.getArg(idx);
+    Value *newArg = newFn->getArg(idx + 2);
+    newArg->setName(oldArg->getName());
+    oldArg->replaceAllUsesWith(newArg);
+  }
+  newFn->getArg(0)->setName("state");
+  newFn->getArg(1)->setName("rcr");
+  return newFn;
+}
+
+/// Return the levels of the caller ray-tracing stages, which will be executed after the function return.
+static unsigned getReturnedLevels(int32_t stage) {
+  // Traversal will return to RGS.
+  if (stage == -1)
+    return 1;
+  rt::RayTracingShaderStage rtStage = static_cast<rt::RayTracingShaderStage>(stage);
+  switch (rtStage) {
+  case rt::RayTracingShaderStage::ShaderStageRayGeneration:
+    llvm_unreachable("Raygen shader should arrive here.\n");
+  case rt::RayTracingShaderStage::ShaderStageClosestHit:
+  case rt::RayTracingShaderStage::ShaderStageMiss:
+    // Traversal | RGS
+    return (1 << 2 | 1);
+  case rt::RayTracingShaderStage::ShaderStageCallable:
+    // CHS/Miss/Callable | Traversal | RGS
+    return (1 << 2 | 1 << 1 | 1);
+  case rt::RayTracingShaderStage::ShaderStageAnyHit:
+    // IS | Traversal | RGS
+    return (1 << 4 | 1 << 2 | 1);
+  default:
+    llvm_unreachable("Unknown raytracing shader type.\n");
+  }
+}
+
+/// Return level of the ray-tracing stage.
+static unsigned getLevelOfRtStage(int32_t stage) {
+  // Traversal
+  if (stage == -1)
+    return 2;
+  rt::RayTracingShaderStage rtStage = static_cast<rt::RayTracingShaderStage>(stage);
+  switch (rtStage) {
+  case rt::RayTracingShaderStage::ShaderStageRayGeneration:
+    return 0;
+  case rt::RayTracingShaderStage::ShaderStageClosestHit:
+  case rt::RayTracingShaderStage::ShaderStageMiss:
+  case rt::RayTracingShaderStage::ShaderStageCallable:
+    return 1;
+  case rt::RayTracingShaderStage::ShaderStageAnyHit:
+    return 3;
+  case rt::RayTracingShaderStage::ShaderStageIntersection:
+    return 4;
+  default:
+    llvm_unreachable("Unknown raytracing shader type.\n");
+  }
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses Continufy::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the Continufy pass \n");
+  m_pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
```

Comment:
Maybe reset `m_pipelineState` at the end of `run`? Otherwise we might have a dangling pointer later on. 

Or even better: Move the implementation of `Continufy` into a separate class `ContinufyPassImpl`, and replace this function by

```c++
ContinufyPassHelper(analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState()).run();
```

Then there is no state that can leak after `run`.

Reply:
I think the pointer member will not be accessed once the `run` end. So I don't think this is a serious issue. Anyway, I have reset it to `nullptr` at the end.

Reply:
Thanks. I agree it's not an issue now. Still I think it is bad style (being aware this pattern is quite common though), and with more complicated passes this becomes more difficult to keep track of.

If we consistently adopted the pattern of using implementation classes for passes, and keeping only state in the pass class that is intended to stay alive, this issue would just go away.

Reply:
I agree resetting member variables at the end of run() is a good idea we should follow, but not sure about having implementation class for each pass. I just realized I don't need the PipelineState. I have removed it.